go/test/endtoend/vtgate/buffer/buffer_test.go

```diff
@@ -0,0 +1,436 @@
+		respByte, _ := ioutil.ReadAll(resp.Body)
+		err := json.Unmarshal(respByte, &resultMap)
+		if err != nil {
+			panic(err)
+		}
+		inFlightMax = getVarFromVtgate(t, label, "BufferLastRequestsInFlightMax", resultMap)
+		masterPromotedCount = getVarFromVtgate(t, label, "HealthcheckMasterPromoted", resultMap)
+		durationMs = getVarFromVtgate(t, label, "BufferFailoverDurationSumMs", resultMap)
+		bufferingStops = getVarFromVtgate(t, "NewMasterSeen", "BufferStops", resultMap)
+	}
+	if inFlightMax == 0 {
+		// Missed buffering is okay when we observed the failover during the
+		// COMMIT (which cannot trigger the buffering).
+		assert.Greater(t, updateThreadInstance.commitErrors, 0, "No buffering took place and the update thread saw no error during COMMIT. But one of it must happen.")
+	} else {
+		assert.Greater(t, inFlightMax, 0)
+	}
+
+	// There was a failover and the HealthCheck module must have seen it.
+	if masterPromotedCount > 0 {
+		assert.Greater(t, masterPromotedCount, 0)
+	}
+
+	if durationMs > 0 {
+		// Number of buffering stops must be equal to the number of seen failovers.
+		assert.Equal(t, masterPromotedCount, bufferingStops)
+	}
+	wg.Wait()
+	clusterInstance.Teardown()
+}
+
+func getVarFromVtgate(t *testing.T, label string, param string, resultMap map[string]interface{}) int {
+	paramVal := 0
+	var err error
+	object := reflect.ValueOf(resultMap[param])
+	if object.Kind() == reflect.Map {
+		for _, key := range object.MapKeys() {
+			if strings.Contains(key.String(), label) {
+				v := object.MapIndex(key)
+				s := fmt.Sprintf("%v", v.Interface())
+				paramVal, err = strconv.Atoi(s)
+				if err != nil {
+					t.Fatal(err.Error())
+				}
+			}
+		}
+	}
+	return paramVal
+}
+
+func externalReparenting(t *testing.T, clusterInstance *cluster.LocalProcessCluster) {
+	start := time.Now()
+
+	// Demote master Query
+	master := clusterInstance.Keyspaces[0].Shards[0].Vttablets[0]
+	replica := clusterInstance.Keyspaces[0].Shards[0].Vttablets[1]
+	master.VttabletProcess.QueryTablet(demoteMasterQuery, keyspaceUnshardedName, true)
+	if master.VttabletProcess.EnableSemiSync {
+		master.VttabletProcess.QueryTablet(disableSemiSyncMasterQuery, keyspaceUnshardedName, true)
+	}
+
+	// Wait for replica to catch up to master.
+	waitForReplicationPos(t, &master, &replica, 60.0)
+
+	duration := time.Since(start)
+	minUnavailabilityInS := 1.0
+	if duration.Seconds() < minUnavailabilityInS {
+		w := minUnavailabilityInS - duration.Seconds()
+		fmt.Printf("Waiting for %.1f seconds because the failover was too fast (took only %.3f seconds)", w, duration.Seconds())
+		time.Sleep(time.Duration(w) * time.Second)
+	}
+
+	// Promote replica to new master.
+	replica.VttabletProcess.QueryTablet(promoteSlaveQuery, keyspaceUnshardedName, true)
+
+	if replica.VttabletProcess.EnableSemiSync {
+		replica.VttabletProcess.QueryTablet(enableSemiSyncMasterQuery, keyspaceUnshardedName, true)
+	}
+	oldMaster := master
+	newMaster := replica
```

Comment:
instead of creating these two variables here, might as well use them from the beginning (line 355)

Reply:
done